compiler/rustc_session/src/cgu_reuse_tracker.rs

```diff
@@ -99,18 +119,24 @@ impl CguReuseTracker {
 
                     if error {
                         let at_least = if at_least { "at least " } else { "" };
-                        let msg = format!(
-                            "CGU-reuse for `{cgu_user_name}` is `{actual_reuse:?}` but \
-                                           should be {at_least}`{expected_reuse:?}`"
-                        );
-                        diag.span_err(error_span.0, &msg);
+                        IncorrectCguReuseType {
+                            span: error_span.0,
+                            cgu_user_name: &cgu_user_name,
+                            actual_reuse,
+                            expected_reuse,
+                            at_least,
+                        };
                     }
                 } else {
+                    //FIXME: Remove this once PR #100694 that implements `[fatal(..)]` is merged
                     let msg = format!(
                         "CGU-reuse for `{cgu_user_name}` (mangled: `{cgu_name}`) was \
                                        not recorded"
                     );
                     diag.span_fatal(error_span.0, &msg)
```

Comment:
This can be changed now that fatal support has landed.

Reply:
Hey @davidtwco,

Thanks for the review :)

I'm not sure I can actually use `emit_fatal` here.

As you can see, this error uses a `diag: &rustc_errors::Handler` param to span the fatal. 

I've tried to implement `fatal` for `rustc_errors::Handler` but I'm wondering what is the best way to implement it?

So far, I've tried to implement it in `[Handler](https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/lib.rs#L522)` like that:

```rust
    pub fn emit_fatal<'a>(&'a self, fatal: impl SessionDiagnostic<'a, !>) -> ! {
// ...
}
```

But it won't wotk because we are at `rustc_session` and `SessionDiagnostic` is not a trait. I guess I cannot use the trait here because it will cause dependency cycles? 

Sp, how one would proceeded to implement this?

Reply:
`check_expected_reuse` is only used in one place, can you just change it to take a `&Session`?

Reply:
Thanks @davidtwco !

Just did the modification, please see the commit [here](https://github.com/rust-lang/rust/pull/100753/commits/71193954f900463c9b00e3485d3430990dce3426).